src/vm/gdbjit.cpp

```diff
@@ -2044,39 +2205,48 @@ bool NotifyGdb::BuildSymbolTableSection(MemBuf& buf, PCODE addr, TADDR codeSize)
     sym[0].st_size = 0;
     sym[0].st_shndx = SHN_UNDEF;
     
-    sym[1].st_name = SymbolNames[1].m_off;
-    sym[1].setBindingAndType(STB_GLOBAL, STT_FUNC);
-    sym[1].st_other = 0;
+    for (int i = 1; i < 1 + method.GetCount(); ++i)
+    {
+        sym[i].st_name = SymbolNames[i].m_off;
+        sym[i].setBindingAndType(STB_GLOBAL, STT_FUNC);
+        sym[i].st_other = 0;
+        sym[i].st_value = SymbolNames[i].m_value - addr;
 #ifdef _TARGET_ARM_
-    sym[1].st_value = 1; // for THUMB code
-#else    
-    sym[1].st_value = 0;
+        sym[i].st_value |= 1; // for THUMB code
 #endif    
-    sym[1].st_shndx = 1; // .text section index
-    sym[1].st_size = codeSize;
+        sym[i].st_shndx = 1; // .text section index
+        sym[i].st_size = SymbolNames[i].m_size;
+    }
 
-    for (int i = 2; i < SymbolCount; ++i)
+    for (int i = 1 + method.GetCount(); i < SymbolCount; ++i)
     {
         sym[i].st_name = SymbolNames[i].m_off;
         sym[i].setBindingAndType(STB_GLOBAL, STT_FUNC);
         sym[i].st_other = 0;
-        sym[i].st_shndx = 11; // .thunks section index
+        sym[i].st_shndx = 11 + (i - (1 + method.GetCount())); // .thunks section index
         sym[i].st_size = 8;
-        sym[i].st_value = SymbolNames[i].m_value;
+        sym[i].st_value = 0;
     }
     return true;
 }
 
 /* Build ELF string section */
 bool NotifyGdb::BuildSectionNameTable(MemBuf& buf)
 {
+    int thunks_count = SymbolCount - 1 - method.GetCount();
     uint32_t totalLength = 0;
     
     /* calculate total size */
     for (int i = 0; i < SectionNamesCount; ++i)
     {
         totalLength += strlen(SectionNames[i]) + 1;
     }
+    for (int i = SectionNamesCount; i < SectionNamesCount + thunks_count; ++i)
```

Comment:
This separation of the code that measures the totalLength and then prints into that seems like a possible source of a problem if in future someone modifies the code by mistake so that the measuring and printing won't exactly match. What I would prefer in this case is to have a loop with two passes when in the first pass, it would use the same code to measure the total length and in the second pass, it would do the actual allocation and writing. That way, it is difficult to make the measurement accidentally different from the copying.
Here is what it could look like:
```C++
buf.MemPtr = nullptr;
for (int pass = 0; pass <= 1; pass++)
{
    size_t totalLength = 0;
    char* bufPtr = buf.MemPtr;
    for (int i = 0; i < SectionNamesCount; ++i)
    {
        size_t length = strlen(SectionNames[i]) + 1;
        if (bufPtr != nullptr)
        {
             strcpy(bufPtr, SectionNames[i]);
             bufPtr += length;
        }
        totalLength += length;
    }
    for (int i = SectionNamesCount; i < SectionNamesCount + thunks_count; ++i)
    {
        char name[256];
        size_t length = sprintf(name, ".thunk_%i", i) + 1;   
        if (bufPtr != nullptr)
        {
             strcpy(bufPtr, name);
             bufPtr += length;
        }
        totalLength += length;
    }

    if (bufPtr == nullptr)
    {
        buf.MemSize = totalLength;
        buf.MemPtr = new (nothrow) char[totalLength];
        if (buf.MemPtr == nullptr)
            return false;
    }
    else 
    {
        ASSERT(buf.MemSize == totalLength);
    }
}

```